cmd/bootstrap/cmd/finalize_test.go

```diff
@@ -0,0 +1,182 @@
+	rootCommit := unittest.StateCommitmentFixture()
+	rootParent := unittest.StateCommitmentFixture()
+	chainName := "main"
+	rootHeight := uint64(1000)
+	epochCounter := uint64(0)
+
+	RunWithSporkBootstrapDir(t, func(bootDir, partnerDir, partnerStakes, internalPrivDir, configPath string) {
+
+		flagConfig = configPath
+		flagPartnerNodeInfoDir = partnerDir
+		flagPartnerStakes = partnerStakes
+		flagInternalNodePrivInfoDir = internalPrivDir
+
+		flagFastKG = true
+
+		flagRootCommit = hex.EncodeToString(rootCommit[:])
+		flagRootParent = hex.EncodeToString(rootParent[:])
+		flagRootChain = chainName
+		flagRootHeight = rootHeight
+		flagEpochCounter = epochCounter
+
+		// set deterministic bootstrapping seed
+		flagBootstrapRandomSeed = deterministicSeed
+
+		hook := zeroLoggerHook{logs: &strings.Builder{}}
+		log = log.Hook(hook)
+
+		finalize(nil, nil)
+		require.Regexp(t, finalizeHappyPathRegex, hook.logs.String())
+		hook.logs.Reset()
+
+		// check if root protocol snapshot exists
+		snapshotPath := filepath.Join(bootDir, model.PathRootProtocolStateSnapshot)
+		assert.FileExists(t, snapshotPath)
+	})
+}
+
+func TestFinalize_Deterministic(t *testing.T) {
+	deterministicSeed := generateRandomSeed()
+	rootCommit := unittest.StateCommitmentFixture()
+	rootParent := unittest.StateCommitmentFixture()
+	chainName := "main"
+	rootHeight := uint64(1000)
+	epochCounter := uint64(0)
+
+	RunWithSporkBootstrapDir(t, func(bootDir, partnerDir, partnerStakes, internalPrivDir, configPath string) {
+
+		flagConfig = configPath
+		flagPartnerNodeInfoDir = partnerDir
+		flagPartnerStakes = partnerStakes
+		flagInternalNodePrivInfoDir = internalPrivDir
+
+		flagFastKG = true
+
+		flagRootCommit = hex.EncodeToString(rootCommit[:])
+		flagRootParent = hex.EncodeToString(rootParent[:])
+		flagRootChain = chainName
+		flagRootHeight = rootHeight
+		flagEpochCounter = epochCounter
+
+		// set deterministic bootstrapping seed
+		flagBootstrapRandomSeed = deterministicSeed
+
+		hook := zeroLoggerHook{logs: &strings.Builder{}}
+		log = log.Hook(hook)
+
+		finalize(nil, nil)
+		require.Regexp(t, finalizeHappyPathRegex, hook.logs.String())
+		hook.logs.Reset()
+
+		// check if root protocol snapshot exists
+		snapshotPath := filepath.Join(bootDir, model.PathRootProtocolStateSnapshot)
+		assert.FileExists(t, snapshotPath)
+
+		// read snapshot
+		firstSnapshot := readRootProtocolSnapshot(t, bootDir)
+
+		// delete snapshot file
+		err := os.Remove(snapshotPath)
+		require.NoError(t, err)
+
+		finalize(nil, nil)
+		require.Regexp(t, finalizeHappyPathRegex, hook.logs.String())
+		hook.logs.Reset()
+
+		// check if root protocol snapshot exists
+		assert.FileExists(t, snapshotPath)
+
+		// read snapshot
+		secondSnapshot := readRootProtocolSnapshot(t, bootDir)
+
+		assert.Equal(t, firstSnapshot, secondSnapshot)
+	})
+}
+
+func RunWithSporkBootstrapDir(t testing.TB, f func(bootDir, partnerDir, partnerStakes, internalPrivDir, configPath string)) {
```

Comment:
👏  Nice, this looks really useful. Would it make sense to move to `utils` as well?

Reply:
Did you mean the `cmd/bootstrap/utils` or `/utils`? I've moved them into `cmd/bootstrap/utils`, because I think the way I have set it up now works well for testing the `finalize` only. For other commands like `keygen` some of the files that is written as part of `RunWithSporkBootstrapDir` is generated by this command, so it still needs some abstractions.

Reply:
I was thinking `/utils` but I agree what you've done makes more sense 👍